### PR TITLE
add argument to allow serving files from additional directories

### PR DESCRIPTION
**Pull request by maxaudron**
# Description

Add the `file_directories` argument to allow serving files from additional directories.
I was working on separating user data from source code for AUTOMATIC1111/stable-diffusion-webui to a directory external to the source and ran into trouble with various files the webui tries to serve from these locations.

# Checklist:

- [X] I have performed a self-review of my own code
- [x] I have added a short summary of my change to the CHANGELOG.md
- [x] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by abidlabs** [comment]
Hi @maxaudron can you elaborate on the issue this PR is fixing a little more?

**Comment by maxaudron** [comment]
@abidlabs I'm using nix to install stable-diffusion-webui as a proper python package, this is stored on a read-only filesystem, but really the same applies for any installed python packages as they should not be directly written to.

stable-diffusion-webui can for example use extensions that provide additional css or javascript files to serve. These are now stored outside the python app source, again, because read-only filesystem of the python app source, and just general good manners.

The source directory is also not necessarily the working directory, causing another potential detachment, I had worked around that by setting the working directory manually to the python app source.

It is a good security consideration to not allow serving from arbitrary directories, but serving files from outside the application's source has plenty of potential uses, and having an escape hatch to allow that is useful.

This way the application can provide a number of directories that are valid to serve files from, like it's own root based on the current file location, or any number of directories provided by configuration.

**Comment by abidlabs** [comment]
Got it, thanks for the clarification. In the future, a better workflow would be to open an issue with the bug / feature request, so that the community and maintaining team can discuss the issue. But as a one off, I'll review this PR.

File: gradio/blocks.py [hunk 1]
```diff
@@ -1258,6 +1260,7 @@ def launch(
         ssl_keyfile_password: str | None = None,
         quiet: bool = False,
         show_api: bool = True,
+        file_directories: List[str] = [],
```
Comment (abidlabs) [inline gradio/blocks.py:1263]: Better to [avoid using a mutable as a default](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/). Can you set this to be `None` and assign to it a list inside of the function?

File: gradio/blocks.py [hunk 2]
```diff
@@ -1287,6 +1290,7 @@ def launch(
             ssl_keyfile_password: If a password is provided, will use this with the ssl certificate for https.
             quiet: If True, suppresses most print statements.
             show_api: If True, shows the api docs in the footer of the app. Default True. If the queue is enabled, then api_open parameter of .queue() will determine if the api docs are shown, independent of the value of show_api.
+            file_directories: List of directories that gradio is allowed to serve files from. Must be absolute paths. Default []
```
Comment (abidlabs) [inline gradio/blocks.py:1293]:
```suggestion
            file_directories: List of directories that gradio is allowed to serve files from in addition to the directory containing the gradio script. Must be absolute paths.
```

File: gradio/routes.py [hunk 3]
```diff
@@ -265,7 +265,12 @@ async def file(path: str, request: fastapi.Request):
             created_by_app = str(utils.abspath(path)) in set().union(
                 *blocks.temp_file_sets
             )
-            if in_app_dir or created_by_app:
+            in_file_dir = any(
+                dir in (str(p) for p in utils.abspath(path).parents)
+                for dir in blocks.file_directories
+            )
```
Comment (abidlabs) [inline gradio/routes.py:271]: Can we do it this way to take advantage of pathlib's built-in `in` functionality
```suggestion
            in_file_dir = any((utils.abspath(dir) in utils.abspath(path).parents for dir in blocks.file_directories))
```

**Comment by abidlabs** [comment]
Hi @maxaudron thanks again for the PR. I made some suggested changes to the PR. In addition, would you please be able to add a unit test that tests this functionality? You can see here for some similar tests: https://github.com/gradio-app/gradio/blob/4aa7d8d74b2d98f7c314e43a902983861d3d4dc9/test/test_routes.py#L183

**Review by abidlabs — CHANGES_REQUESTED** [review]
See above

**Comment by abidlabs** [comment]
@maxaudron is this ready to be re-reviewed?

**Comment by maxaudron** [comment]
@abidlabs yes, I've applied your suggestions and added a test

**Review by abidlabs — APPROVED** [review]

**Comment by abidlabs** [comment]
Made a tweak to the test code, will merge in after it passes, thanks for the nice PR @maxaudron!

**Pull request merged**